### DISCUSSION TITLE
Using in Windows + SFTP

### DISCUSSION THREAD

**Discussion by cristian-spiescu** [netinvent/npbackup#36] [Q&A] [answered]
Hi there!

I found this piece of software = npbackup, which seems really interesting. I'm trying to use it to replace my current `.bat` files. I have 2 questions

1/ I assume many options from the GUI map to restic command line options. Is there a way to see the command line that is generated based on the input from the GUI? It may help.

2/ Using from windows + FTP: I [understand](https://forum.restic.net/t/tutorial-backup-from-windows-with-putty-plink-to-a-sftp-server/2489) it doesn't work out of the box. Hence my command line looks like this:

`restic -o sftp.command="plink -s -l ??? -pw ??? -P ??? myserver.com sftp" -r sftp:host:/some-path -p password.txt ...`

What are the corresponding settings for this in npbackup?

Thanks!

**Comment by deajan** [comment]
1. Indeed. Run with `--debug` to see what actually happens.
2. That's what `additional_parameters` is supposed to solve

Keep me posted if you need more control on parameters.

**Comment by cristian-spiescu** [comment]
Thanks for the info :)

1/ I don't understand. I ran w/ --debug, and I have this output:

```
2023-11-18 20:54:59,847 :: DEBUG :: Command ""C:\Users\user\AppData\Local\Temp\onefile_30172_133448107262900131\npbackup\restic.exe" init --repository-version 2 --compression auto" returned with exit code "1". Command output was:
2023-11-18 20:54:59,847 :: DEBUG :: STDOUT: subprocess ssh: ssh: Could not resolve hostname host: No such host is known.

Fatal: create repository at sftp:host:/bla-bla failed: unable to start the sftp session, error: EOF
```
The command line of the `restic` executable is just ` init --repository-version 2 --compression auto`, cf. output. But this cannot be true. The command complains about not being able to find the URL. So how does restic know the URL? I would have expected to see in the command line: `-r sftp:...`.

**Comment by cristian-spiescu** [comment]
2/ adding something in `additional_parameters` => I don't see them neither in the console output

**Comment by cristian-spiescu** [comment]
Any idea what I'm doing wrong, cf. my replies above?

**Comment by deajan** [comment]
Sorry, yes I totally see what the problem is.
Additionnal parameters are given to backup execution, but not to init step.
Do you run the pre-built executables or the python version ?
I can make a quick fix for the python version if you can test it, than I can rebuild the executables.

**Comment by cristian-spiescu** [comment]
I am currently using the executable version, in Win. But I can switch to python mode. I guess I need to:

* install python
* pip install npbackup

Regarding:

> tkinter via yum install python-tkinter or apt install python3-tk.

I'm under the impression that the python installer proposes them already.

**Comment by deajan** [comment]
You won't need to install tkinter on Windows since the official builds already include it ;)
Pretty straightforward, download latest python from python.org, install to let's say c:\python311 and decompress npbackup sources to let's say c:\npbackup

Then
```
cd c:\NPBACKUP
c:\python311\python.exe -m venv venv
c:\NPBACKUP\venv\scripts\python.exe -m pip install -r c:\npbackup\requirements.txt
SET PYTHONPATH=c:\npbackup
c:\NPBACKUP\venv\scripts\python.exe bin\npbackup
```

I can make a fix until tomorrow for this on a separate git branch.

**Comment by cristian-spiescu** [comment]
Thanks. I cloned the repo, and performed the instructions. The program complained that it didn't found `restic`. It appears that it needs to see the `restic.exe` file in one among some specific locations. In my case, it was installed via `scoop`. So it was available in `PATH`, but not there where `npbackup` wanted it.

So I downloaded the `restic.exe` binary and copied it. At version 0.16.2. Is this OK? Or should I download a specific version?

> I can make a fix until tomorrow for this on a separate git branch.

Super. I'll give it a try asap.

Next, if everything works, I'm thinking in maybe contributing somehow our stuff that we have in our scripts:
* make snapshot for LVM volumes
* send message to healthchecks.io

**Comment by deajan** [comment]
Any restic version from 0.13 to 0.16.2 will do.
I also have a wrapper script for npbackup which handles qcow2 disk snapshots that I can share if needed.
How about healthchecks.io ? Aren't the prometheus metrics sufficient ?

PS: currently giving some love to another of my repos, I will probably have a fix up at the end of the weekend ;)

**Comment by cristian-spiescu** [comment]
> How about healthchecks.io ? Aren't the prometheus metrics sufficient ?

For more "consumer" type users, I find this easier to use. Using prometheus, requires a "pro-user", that has already setup a prometheus infrastructure and associated processes. We have this for our apps. But for the backup, I started w/ healthchecks for my personal laptop, and I didn't want to migrate.

> PS: currently giving some love to another of my repos, I will probably have a fix up at the end of the weekend ;)

Sure; no problem.

**Comment by deajan** [comment]
@cristian-spiescu I've uploaded branch urgent-fixes with a couple of hotfixes, including the one I intend for the SFTP usecase, which was introduced trying to resolve another bug. Mind giving it a try ?

**Comment by deajan** [comment]
Any updates on this ?

**Reply by cristian-spiescu** [reply]
I'm really sorry. Somehow I missed the notifications. 

Thanks a lot for the commit.

I think I can find some time this week.

**Comment by deajan** [comment]
No problem.
Anyway, I'm mostly at the end of a major rewrite of NPBackup, that changes all the under the hood management, as well as overall UX.
If you have any issues with current implementation, let me know so I can get next one right ;)

**Comment by cristian-spiescu** [comment, marked as answer]
Hi @deajan,

W/ a bit of delay 🙂, I restarted the project of using npbackup. It seems that since [quite a while](https://forum.restic.net/t/how-to-set-up-smb-or-sftp-on-windows/8316/6) there is default ssh support. Leveraging that, I could connect to the sftp repo w/o using additional params (plink ...).

I'd have a small piece of advice. I'm observing that this discussion post (and others as well) are closed. Actually, I didn't even know that discussions are closable (like issues). I would suggest not to close the discussions. Searching will be easier for people.

**Comment by deajan** [comment]
@cristian-spiescu Sounds like a good piece of advice to me.
Feel free to open other issues / discussions if needed.
